fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionGraph.java

```diff
@@ -0,0 +1,125 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+package com.starrocks.transaction;
+
+import com.google.common.base.Preconditions;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class DatabaseTransactionGraph {
+    private static final Logger LOG = LogManager.getLogger(DatabaseTransactionGraph.class);
+
+    static class Node {
+        long txnId;
+        List<Long> writeTableIds;
+        // transactions this txn depends
+        Set<Node> ins;
+        // transactions depending on this txn
+        Set<Node> outs;
+
+        Node(long txnId, List<Long> writeTableIds) {
+            this.txnId = txnId;
+            this.writeTableIds = writeTableIds;
+        }
+
+        void addIns(Node in) {
+            if (ins == null) {
+                ins = new HashSet<>();
+            }
+            ins.add(in);
+        }
+
+        void addOuts(Node out) {
+            if (outs == null) {
+                outs = new HashSet<>();
+            }
+            outs.add(out);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Node node = (Node) o;
+            return txnId == node.txnId;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(txnId);
+        }
+    }
+
+    private Map<Long, Node> nodes = new HashMap<>();
+    private Set<Node> nodesWithoutIns = new HashSet<>();
+
+    // tableid -> txnId that lastly write this table
+    private Map<Long, Node> lastTableWriter = new HashMap<>();
+
+    public DatabaseTransactionGraph() {
+    }
+
+    public int size() {
+        return nodes.size();
+    }
+
+    public void add(long txnId, List<Long> writeTableIds) {
+        if (nodes.containsKey(txnId)) {
+            LOG.warn("add an already exist txn:{}", txnId);
+            return;
+        }
+        Node node = new Node(txnId, writeTableIds);
+        for (long tableId : writeTableIds) {
+            Node previous = lastTableWriter.put(tableId, node);
+            if (previous != null) {
+                Preconditions.checkState(previous != node, "duplicate node {}", txnId);
+                node.addIns(previous);
```

Comment:
Does this mean that we decide txn dependency by committed order not by txn start time(i.e. txn id)?

Reply:
yes